### PR TITLE
Multi head attention support

### PR DESCRIPTION
**Pull request by elad-c**
Support MultiHeadAttention layer in Tensorflow framework:

substitute the layer with an implementation of Dense, Reshape, Permute, Concat & Dot layers.

note: The keras layer allows running a loop on certain axes of the input tensor\s. In the current substitution axes that are looped on are folded to the bacth axis.

### CODE REVIEW

File: model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+        self.k_perm_shape = tuple([self.key_shape[i] for i in self.perm_dims])
+        self.v_perm_shape = tuple([self.value_shape[i] for i in self.perm_dims])
+        self.q_reshape_shape = tuple([self.query_shape[0], self.q_att_axes_prod, self.query_shape[-1]])
+        self.k_reshape_shape = tuple([self.key_shape[0], kv_att_axes_prod, self.key_shape[-1]])
+        self.v_reshape_shape = tuple([self.value_shape[0], kv_att_axes_prod, self.value_shape[-1]])
+        self.query_proj_shape = self.q_reshape_shape[:-1] + (self.query_key_dim,)
+        self.key_proj_shape = self.k_reshape_shape[:-1] + (self.query_key_dim,)
+        self.value_proj_shape = self.v_reshape_shape[:-1] + (self.value_dim,)
+        self.att_matrix_shape = (self.key_shape[0],) + (self.q_att_axes_prod, kv_att_axes_prod)
+        self.reuse_params = {'reuse': mha_node.reuse, 'reuse_group': mha_node.reuse_group}
```
Comment (haihabi) [inline model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py:74]: Move the string to constants

File: model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        :param k_reshape_node:
+        :param v_reshape_node:
+        :param mha_node: MHA node
+        :param head_index: head index being generated. used to set correct node names
+        :param params: MHA params object
+        :return: output of attention head
+        """
+        # add norm factor to query kernel and bias
+        factor = (params.query_key_dim ** -0.5)
+        qk = mha_node.weights[self._get_weight_by_name(mha_node, '/query/kernel')][:, head_index, :].copy() * factor
```
Comment (haihabi) [inline model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py:160]: String to constant

File: model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+
+        # Generate nodes for attention heads:
+        att_head_output_nodes = []
+        for head_index in range(params.num_heads):
+            dotv_node = self._calc_attention_head(graph, q_reshape_node, k_reshape_node, v_reshape_node,
+                                                  mha_node, head_index, params)
+            att_head_output_nodes.append(dotv_node)
+
+        # concatenate all attention heads
+        concat_node = BaseNode(f'{mha_node.name}_concat', {},
```
Comment (haihabi) [inline model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py:263]: Make this a function

File: model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+        graph.add_node_with_in_edges(q_node, [q_reshape_node])
+        k_node = BaseNode(f'{mha_node.name}_key_{head_index}', {UNITS: params.query_key_dim, USE_BIAS: params.use_bias, ACTIVATION: LINEAR},
+                          params.k_reshape_shape, params.key_proj_shape, {KERNEL: kk, BIAS: kb}, Dense, **params.reuse_params)
+        # key_nodes.append(k_node)
+        # graph.add_node(k_node)
+        graph.add_node_with_in_edges(k_node, [k_reshape_node])
+        v_node = BaseNode(f'{mha_node.name}_value_{head_index}', {UNITS: params.value_dim, USE_BIAS: params.use_bias, ACTIVATION: LINEAR},
+                          params.v_reshape_shape, params.value_proj_shape, {KERNEL: vk, BIAS: vb}, Dense, **params.reuse_params)
+        # value_nodes.append(v_node)
+        # graph.add_node(v_node)
```
Comment (haihabi) [inline model_compression_toolkit/keras/graph_substitutions/substitutions/multi_head_attention_decomposition.py:183]: Remove irrelevant comments
